SARAALERT-1421: Improve loading of assessments table on patient page

timwongj opened the pull request:
# Description
Jira Ticket: [SARAALERT-1421](https://tracker.codev.mitre.org/browse/SARAALERT-1421)

Currently, hundreds of queries (although cached, still slower) are being executed to load the assessments table, which is responsible for a noticeable amount of the loading time of the patient page. This is due to the fact that each symptom of each assessment is being queried separately.

The changes in this PR aim to improve the performance of this by querying these assessments and symptoms in bulk rather than individually

# Checklists

**Submitter:**
- [x] This PR describes why these changes were made.
- [x] This PR is into the correct branch.
- [x]  This PR includes the correct JIRA ticket reference.
- [x] Code diff has been reviewed (it **does not** contain: additional white space, not applicable code changes, debug statements, etc.)
- [x] **N/A** If UI changes have been made, Chrome Dev Tools Lighthouse accessibility test has been executed to ensure no 508 issues were introduced.
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass (remember to update Gemfile when applicable)
- [x] Test fixtures updated and documented as necessary


@jkufro  :
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code
- [x] If applicable, you have tested changes against a large database, and considered possible performance regressions


@mayerm94  :
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code
- [x] If applicable, you have tested changes against a large database, and considered possible performance regressions


@holmesie  :
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code
- [x] If applicable, you have tested changes against a large database, and considered possible performance regressions

jkufro commented on app/controllers/patients_controller.rb:
```diff
@@ -39,9 +39,6 @@ def show
     @history_types = History::HISTORY_TYPES
-
-    # If we failed to find a subject given the id, redirect to index
-    redirect_to(root_url) && return if @patient.nil?
```
Removing this is good since this check is already done on L22

jkufro commented on app/helpers/assessment_query_helper.rb:
```diff
@@ -34,7 +34,7 @@ def sort(assessments, order, direction)
       assessments = assessments.order(created_at: dir)
     else
       # Verify this is a sort request for a symptom column.
-      symptom_columns = Assessment.get_unique_symptoms_for_assessments(assessments.pluck(:id)).pluck(:name)
```
Why not improve the performance of the `Assessment.get_unique_symptoms_for_assessments` method so that it can be reused elsewhere?

timwongj replied:
great question, so this method was originally being used in 2 places and I started out by optimizing it for both, but the more I continued to optimize it, it seemed like it would be better to just optimize them individually because they were being used for different purposes/returning different fields

jkufro commented:
@timwongj would it be possible to add a benchmark that shows the performance gain with this PR? If that's too much for this, then do you have measurements elsewhere that you took on the performance improvement?

Could add a micro benchmark like this so it could run in that same job
```ruby
patient = Patient.where(monitoring: true).first
micro_results << benchmark(
  name: 'PatientAssessmentQuerying',
  time_threshold: 1,
  no_exit: true,
  setup: proc {
    patient.do_something_to_add_lots_of_assessments
  }
) {
  assessments = search(patient.assessments, search_text)
  assessments = sort(assessments, sort_order, sort_direction)
  assessments = paginate(assessments, entries, page)
  assessments = format_for_frontend(assessments)
}
```

jkufro commented on app/helpers/assessment_query_helper.rb:
```diff
@@ -69,35 +69,77 @@ def paginate(assessments, entries, page)
+  end
+
+  # Gets all unique symptoms (based on name and threshold_condition_hash) for a given array of assessment IDs.
+  def get_unique_threshold_symptoms(assessment_ids)
```
where is this method used?

timwongj replied:
ah, this was originally renamed from `get_unique_symptoms_for_assessments` but is no longer being used, so I will remove it

jkufro replied:
Ok feel free to resolve this once you push that change

jkufro commented on app/helpers/assessment_query_helper.rb:
```diff
@@ -69,35 +69,77 @@ def paginate(assessments, entries, page)
 
   # Formats assessments to be displayed on the frontend.
   def format_for_frontend(assessments)
```
I'm seeing 5 different SQL queries being sent to the database in this method. Is there a way that we could refactor this to use fewer DB calls with joins, eager_loads, etc?

timwongj replied:
so after spending a couple hours playing around with this, I think the way it's written now is better than trying to do any kind of joins because that seems to be slower. I tried doing stuff like this to get conditions and symptoms at the same time, but it's either about the same or slower in terms of speed:

```
threshold_symptoms = Symptom.joins('JOIN conditions reported_conditions ON symptoms.condition_id = reported_conditions.id')
                            .joins('JOIN conditions threshold_conditions ON threshold_conditions.threshold_condition_hash = reported_conditions.threshold_condition_hash')
                            .joins('JOIN assessments ON reported_conditions.assessment_id = assessments.id')
                            .where('assessments.id IN (?)', assessment_ids)
                            .where('threshold_conditions.type = ?', 'ThresholdCondition')
                            .select('threshold_conditions.threshold_condition_hash', 'symptoms.name', 'symptoms.label', 'symptoms.type', 'symptoms.required', 'symptoms.threshold_operator', 'symptoms.bool_value', 'symptoms.int_value', 'symptoms.float_value')
                            .group_by(&:threshold_condition_hash)
                            .transform_values { |symptoms| Hash[symptoms.map { |symptom| [symptom[:name], symptom] }] }
```

Before:
```
(0.5ms)  SELECT `conditions`.`id`, `conditions`.`threshold_condition_hash` FROM `conditions` WHERE `conditions`.`type` = 'ThresholdCondition' AND `conditions`.`threshold_condition_hash` IN (SELECT `conditions`.`threshold_condition_hash` FROM `conditions` WHERE `conditions`.`type` = 'ReportedCondition' AND `conditions`.`assessment_id` IN (2691, 1592, 1443, 747, 172, 134, 49))
  ↳ app/helpers/assessment_query_helper.rb:81:in `format_for_frontend'
Symptom Load (0.4ms)  SELECT `symptoms`.`condition_id`, `symptoms`.`name`, `symptoms`.`label`, `symptoms`.`type`, `symptoms`.`required`, `symptoms`.`threshold_operator`, `symptoms`.`bool_value`, `symptoms`.`int_value`, `symptoms`.`float_value` FROM `symptoms` WHERE `symptoms`.`condition_id` IN (13, 8)
  ↳ app/helpers/assessment_query_helper.rb:86:in `format_for_frontend'
```

After:
```
Symptom Load (1.3ms)  SELECT `threshold_conditions`.`threshold_condition_hash`, `symptoms`.`name`, `symptoms`.`label`, `symptoms`.`type`, `symptoms`.`required`, `symptoms`.`threshold_operator`, `symptoms`.`bool_value`, `symptoms`.`int_value`, `symptoms`.`float_value` FROM `symptoms` JOIN conditions reported_conditions ON symptoms.condition_id = reported_conditions.id JOIN conditions threshold_conditions ON threshold_conditions.threshold_condition_hash = reported_conditions.threshold_condition_hash JOIN assessments ON reported_conditions.assessment_id = assessments.id WHERE (assessments.id IN (2691,1592,1443,747,172,134,49)) AND (threshold_conditions.type = 'ThresholdCondition')
  ↳ app/helpers/assessment_query_helper.rb:97:in `group_by'
```

jkufro reviewed (changes requested):
Haven't looked at JS changes yet but have some questions about the Rails side before I move on to that part

jkufro commented on app/helpers/assessment_query_helper.rb:
```diff
@@ -61,43 +61,76 @@ def sort(assessments, order, direction)
-  def format_for_frontend(assessments)
-    table_data = []
-    symptoms = Assessment.get_unique_symptoms_for_assessments(assessments.pluck(:id))
+  def self.format_for_frontend(assessments)
```
I was looking deeper into the micro benchmark that was made and there seems to be something strange going on with the `format_for_frontend` method. It seems to consistently give blank results (as if no assessments were passed to the function. 

## Right before the assessments loop
![Screen Shot 2021-04-20 at 6 38 18 AM](https://user-images.githubusercontent.com/24628687/115397140-ff9a2c80-a1a2-11eb-9708-94245c95707a.png)

## What is being returned by the function
![Screen Shot 2021-04-20 at 6 38 49 AM](https://user-images.githubusercontent.com/24628687/115397196-12acfc80-a1a3-11eb-99e1-9d9c7cfe5e49.png)

jkufro replied:
putting a byebug at the top of the method shows that assessments seem to still be intact as they are passed to `format_for_frontend`

![Screen Shot 2021-04-20 at 6 40 31 AM](https://user-images.githubusercontent.com/24628687/115397417-4ee05d00-a1a3-11eb-9c58-cb8a0f88d343.png)

jkufro replied:
The root of the issue appears to be in the `assessment_ids = assessments.map(&:id)` line

jkufro replied:
@timwongj changing the `paginate` page argument to 0  fixes this. It looks like that method was written to always add `1` to the page number that passed - so probably expecting page 1 to be index 0. Page 2 in this case would definitely have 0 entries since all 100 are shown on page 1

timwongj replied:
made the change here: https://github.com/SaraAlert/SaraAlert/pull/806/commits/c593b41a33decfcecc95ce7cb6d875fadcc59129#diff-eb2f94faa1282056a764ceeae2b4bb03b361611302fd04d7b0ba3b3949573ec4R66

thanks for debugging and catching this!

jkufro commented on performance/benchmarks/micro_benchmarks.rb:
```diff
@@ -31,6 +32,43 @@ def divider
+patient = Patient.create!(creator: User.first, jurisdiction: Jurisdiction.first, responder: Patient.first)
+micro_results << benchmark(
+  name: 'PatientAssessmentQuerying',
+  time_threshold: 1,
```
May want to decrease this further, but let's see how it performs near the end of the PR before changing it.

timwongj replied:
fixed here: https://github.com/SaraAlert/SaraAlert/pull/806/commits/c593b41a33decfcecc95ce7cb6d875fadcc59129#diff-eb2f94faa1282056a764ceeae2b4bb03b361611302fd04d7b0ba3b3949573ec4R40

jkufro commented on performance/benchmarks/micro_benchmarks.rb:
```diff
@@ -31,6 +32,43 @@ def divider
 
 divider
 
+# Time Threshold: 1 second
```
My data from quickly running locally:

## This PR (after pagination change)
- 0.150s
- 0.161s
- 0.153s

## 1.28.0
- 3.353s
- 3.390s
- 3.300s

timwongj replied:
thanks for running some benchmarks on it, I ran it locally too and it seemed to be in the low 0.1s for me, so I set the threshold to 0.2s for now

jkufro commented on app/controllers/assessments_controller.rb:
```diff
@@ -21,10 +19,10 @@ def index
+    assessments = AssessmentQueryHelper.search(assessments, search_text)
+    assessments = AssessmentQueryHelper.sort(assessments, sort_order, sort_direction)
+    assessments = AssessmentQueryHelper.paginate(assessments, entries, page)
+    assessments = AssessmentQueryHelper.format_for_frontend(assessments)
```
Why don't we have a test file for `AssessmentQueryHelper`'s methods?

timwongj replied:
seems like tests were just never written for the assessment query helper yet. I would offer to do it, but I think that might be outside the scope of this ticket (other than the `format_for_frontend` method since this is more focused on the assessment table loading performance). I have included system tests in this PR that verify each assessment, its symptom values, the symptomatic styling/color of each cell and row, etc, which I think is sufficient enough for now to verify the correctness of the code being changed

jkufro reviewed (changes requested):
left some additional comments that should be addressed

jkufro reviewed (changes requested):
All of my comments from before have been resolved. 

I'm seeing test failures in CI with the AsessmentsController (and the same errors locally as well). 

Also, is this branch up to date with 1.28.0?

timwongj commented:
> All of my comments from before have been resolved.
> 
> I'm seeing test failures in CI with the AsessmentsController (and the same errors locally as well). Looks like an issue with `Assessment.rb:L52`
> 
> Also, is this branch up to date with 1.28.0?

just rebased and fixed the tests: https://github.com/SaraAlert/SaraAlert/pull/806/commits/d57204df32602bccc846366a71ff9ce2c7b28bb8

thanks @jkufro !

jkufro commented on app/javascript/components/layout/CustomTable.js:
```diff
@@ -324,6 +333,7 @@ class CustomTable extends React.Component {
+  dataType: PropTypes.string.required,
   columnData: PropTypes.array,
   rowData: PropTypes.array,
   totalRows: PropTypes.number,
```
Not sure I understand enough about the JS side of this to know why this is happening. Is there a way to resolve this? @timwongj 

![Screen Shot 2021-04-21 at 12 35 01 PM](https://user-images.githubusercontent.com/24628687/115603930-ff328c00-a29d-11eb-8205-9b471a4329e5.png)

jkufro replied:
This error happens for me on a patient show page

timwongj replied:
great catch, fixed here: https://github.com/SaraAlert/SaraAlert/pull/806/commits/79c99d1bdff2281ff7c2e159842e8936ae6c9ec3

thanks @jkufro

jkufro reviewed (approved):
Thank you @timwongj for addressing all feedback
